crates/astria-sequencer/src/slinky/oracle/state_ext.rs

```diff
@@ -0,0 +1,278 @@
+            None => Ok(None),
+        }
+    }
+
+    #[instrument(skip_all)]
+    async fn get_currency_pair_mapping(&self) -> Result<HashMap<u64, CurrencyPair>> {
+        let prefix = format!("{CURRENCY_PAIR_TO_ID_PREFIX}/");
+        let mut currency_pairs = HashMap::new();
+
+        let mut stream = std::pin::pin!(self.prefix_keys(&prefix));
+        while let Some(Ok(key)) = stream.next().await {
+            let Some(bytes) = self
+                .get_raw(&key)
+                .await
+                .context("failed reading currency pair id from state")?
+            else {
+                bail!("currency pair not found in state; this is a bug")
+            };
+            let Id(id) = Id::try_from_slice(&bytes).context("invalid currency pair id bytes")?;
+
+            let currency_pair = key
+                .strip_prefix(&prefix)
+                .context("failed to strip prefix from currency pair state key")?
+                .parse::<CurrencyPair>()
+                .context("failed to parse storage key suffix as currency pair")?;
+            currency_pairs.insert(id, currency_pair);
+        }
+        Ok(currency_pairs)
+    }
+
+    #[instrument(skip_all)]
+    async fn get_num_currency_pairs(&self) -> Result<u64> {
+        let Some(bytes) = self
+            .get_raw(NUM_CURRENCY_PAIRS_KEY)
+            .await
+            .context("failed reading number of currency pairs from state")?
+        else {
+            return Ok(0);
+        };
+        let Count(num_currency_pairs) =
+            Count::try_from_slice(&bytes).context("invalid number of currency pairs bytes")?;
+        Ok(num_currency_pairs)
+    }
+
+    #[instrument(skip_all)]
+    async fn get_num_removed_currency_pairs(&self) -> Result<u64> {
+        let Some(bytes) = self
+            .get_raw(NUM_REMOVED_CURRENCY_PAIRS_KEY)
+            .await
+            .context("failed reading number of removed currency pairs from state")?
+        else {
+            return Ok(0);
+        };
+        let Count(num_removed_currency_pairs) = Count::try_from_slice(&bytes)
+            .context("invalid number of removed currency pairs bytes")?;
+        Ok(num_removed_currency_pairs)
+    }
+
+    #[instrument(skip_all)]
+    async fn get_currency_pair_state(
+        &self,
+        currency_pair: &CurrencyPair,
+    ) -> Result<Option<CurrencyPairState>> {
+        let bytes = self
+            .get_raw(&currency_pair_state_storage_key(currency_pair))
+            .await
+            .context("failed to get currency pair state from state")?;
+        match bytes {
+            Some(bytes) => {
+                let currency_pair_state = serde_json::from_slice(&bytes)
+                    .context("failed to deserialize currency pair state")?;
+                Ok(Some(currency_pair_state))
+            }
+            None => Ok(None),
+        }
+    }
+
+    #[instrument(skip_all)]
+    async fn get_all_currency_pairs(&self) -> Result<Vec<CurrencyPair>> {
```

Comment:
As above, implement this as a stream and let the call site decide what to do with this. See https://github.com/astriaorg/astria/pull/1453 for how to do that.